OpenTTD/OpenTTD
# Feature: play easier together with friends from behind home routers

Integrated #9234. Closes #9234.

~Test binaries can be found here: https://www.openttd.org/downloads/openttd-branches/pr9017/latest.html
See below how to help out testing.~

~If you build yourself, make sure to name your branch "pr9017". Otherwise your revision will not match anyone using the binaries above.~

**!!NOTE!!** Currently the Game Coordinator is not working with this PR, while I rebuild this PR from the ground up. See https://gist.github.com/TrueBrain/75c016cdb90efa18cb79f3f6bae9af25 for details.

## Motivation / Problem

Currently, to host a network server you really need to know a few things about routers and networking. This is way to advanced for the average Joe, and this becomes very clear when looking at Steam Discussions.
For Rich Presence (Discord, Steam, ..) we need "invite codes" so people can join each others game.

So, can we do this? I wrote a gist to explain the process of getting to this PR: https://gist.github.com/TrueBrain/f5685dc0fdbd012aac48ea2a36158198

But in case you are not into reading, let me summarize:

- Every invite-only / public server creates a persistent TCP connection to a Game Coordinator (replacing the master-server)
- Every server gets assigned an unique "join key" (called an "invite code" in the UI)
- Server listing returns only the "join key" of a server
- A client can join any server based on this "join key" (without even knowing the IP of the server), either via listing or directly inputting it
- The Game Coordinator will try to connect the server and client. It tries 3 methods: direct connect, STUN, and if all fails, TURN (see gists for details)

There are now three kind of servers:
- Local servers: do not connect to the Game Coordinator. "Search LAN" does work for these servers, as do people with the public IP and if port-forwarding is setup correctly. Just the old "non-advertised" servers.
- Invite-only servers: the owner of the server needs to give the invite-code to someone before he can join.
- Public servers are the old advertised servers.

## Description

See the individual commits and the gist above. Not much more to add.

## Limitations

- Server info is delayed by 30s (by design). You can refresh individual servers if you want up-to-date information.

## Testing

If you want to help out testing, great!

- Start the client by downloading it here: https://www.openttd.org/downloads/openttd-branches/pr9017/latest.html (or build it yourself, make sure you name your branch `pr9017`!)
- Start a server, make it public or invite-only
- Ask someone to join your server

Notice how you didn't have to setup firewalls, port-forwarding, etc?

Let me know if you find any problems, etc.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)
